src/python/pants/backend/go/util_rules/third_party_pkg.py

```diff
@@ -90,118 +96,360 @@ class AllThirdPartyPackagesRequest:
-        go_mod_digest = await Get(Digest, AddPrefix(request.go_mod_digest, go_mod_dir))
-    else:
-        go_mod_digest = request.go_mod_digest
-
-    list_argv = (
-        "list",
-        # This rule can't modify `go.mod` and `go.sum` as it would require mutating the workspace.
-        # Instead, we expect them to be well-formed already.
-        #
-        # It would be convenient to set `-mod=mod` to allow edits, and then compare the resulting
-        # files to the input so that we could print a diff for the user to know how to update. But
-        # `-mod=mod` results in more packages being downloaded and added to `go.mod` than is
-        # actually necessary.
-        # TODO: nice error when `go.mod` and `go.sum` would need to change. Right now, it's a
-        #  message from Go and won't be intuitive for Pants users what to do.
-        "-mod=readonly",
-        # There may be some packages in the transitive closure that cannot be built, but we should
-        # not blow up Pants.
-        #
-        # For example, a package that sets the special value `package documentation` and has no
-        # source files would naively error due to `build constraints exclude all Go files`, even
-        # though we should not error on that package.
-        "-e",
-        "-json",
-        # This matches all packages. `all` only matches first-party packages and complains that
-        # there are no `.go` files.
-        "...",
+@dataclass(frozen=True)
+class ModuleDescriptorsRequest:
+    digest: Digest
+    path: str
+
+
+@dataclass(frozen=True)
+class ModuleDescriptor:
+    import_path: str
+    name: str
+    version: str
+    indirect: bool
+    minimum_go_version: str | None
+    sum: str | None
+
+
+class ModuleDescriptors(DeduplicatedCollection[ModuleDescriptor]):
+    """An ordered list of `ModuleDescriptor`s."""
+
+
+@dataclass(frozen=True)
+class AnalyzeThirdPartyModuleRequest:
+    go_mod_digest: Digest
+    go_mod_path: str
+    import_path: str
+    name: str
+    version: str
+    minimum_go_version: str | None
+
+
+@dataclass(frozen=True)
+class AnalyzedThirdPartyModule:
+    packages: FrozenOrderedSet[ThirdPartyPkgAnalysis]
+
+
+@dataclass(frozen=True)
+class AnalyzeThirdPartyPackageRequest:
+    pkg_json: FrozenDict[str, Any]
+    module_sources_digest: Digest
+    module_sources_path: str
+    module_import_path: str
+    package_path: str
+    minimum_go_version: str | None
+
+
+@dataclass(frozen=True)
+class FallibleThirdPartyPkgAnalysis:
+    """Metadata for a third-party Go package, but fallible if our analysis failed."""
+
+    analysis: ThirdPartyPkgAnalysis | None
+    import_path: str
+    exit_code: int = 0
+    stderr: str | None = None
+
+
+@rule
+async def analyze_module_dependencies(request: ModuleDescriptorsRequest) -> ModuleDescriptors:
+    mod_list_result = await Get(
+        ProcessResult,
+        GoSdkProcess(
+            command=["list", "-mod=readonly", "-e", "-m", "-json", "all"],
```

Comment:
Imo it would be good to restore the comments for each arg.

Reply:
Done.